walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,65 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Comparator;
+import java.util.Arrays;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+      Collection<Event> eventsAL = new ArrayList<Event>(events);
+      Collection<String> people = request.getAttendees();
+
+      Collection<TimeRange> availableTimes = new ArrayList<TimeRange>();
+      availableTimes.add(TimeRange.WHOLE_DAY);
+      getAvailableTimes(eventsAL, people, availableTimes, request.getDuration());
+    
+      Collection<TimeRange> availableTimesOptional = new ArrayList<TimeRange>(availableTimes);
+      Collection<String> optionalPeople = request.getOptionalAttendees();
+      getAvailableTimes(eventsAL, optionalPeople, availableTimesOptional, request.getDuration());
+
+      if(request.getAttendees().isEmpty() && availableTimesOptional.isEmpty()){
+          return Arrays.asList();
+      }
+
+      return availableTimesOptional.isEmpty() ? availableTimes : availableTimesOptional;
+  }
+
+  private void getAvailableTimes(Collection<Event> events, Collection<String> people, Collection<TimeRange> availableTimes, long duration){
+      for(Event event : new ArrayList<Event>(events)){
+          if(!Collections.disjoint(event.getAttendees(), people)){
+            createAvailableRange(availableTimes, event.getWhen());
+            events.remove(event);
+          }
+      }
+      availableTimes.removeIf(item -> (item.duration() < duration));
+      return;
+  }
+
+  private void createAvailableRange(Collection<TimeRange> availableTimes, TimeRange newEvent){
+      for(TimeRange event : new ArrayList<TimeRange>(availableTimes)){
```

Comment:
same comment as above

Reply:
when I try to remove by gathering, I get a concurrentModificationException
`private void createAvailableRange(Collection<TimeRange> availableTimes, TimeRange newEvent){
      List<TimeRange> deleteList = new ArrayList<TimeRange>();
      for(TimeRange event : availableTimes){
        if(event.contains(newEvent)){
            deleteList.add(event);
            availableTimes.add(TimeRange.fromStartEnd(event.start(), newEvent.start(), false));
            availableTimes.add(TimeRange.fromStartEnd(newEvent.end(), event.end(), false));
            break;
        } else if(event.equals(newEvent)){
            deleteList.add(event);
            break;
        }  else if(newEvent.contains(event)){
            deleteList.add(event);
        } else if(event.overlaps(newEvent)){
            if(event.start() < newEvent.end()){
                deleteList.add(event);
                availableTimes.add(TimeRange.fromStartEnd(newEvent.end(), event.end(), false));
            } else if(newEvent.start() < event.end()){
                deleteList.add(event);
                availableTimes.add(TimeRange.fromStartEnd(event.start(), newEvent.start(), false));
            }
        }
      }
      availableTimes.removeAll(deleteList);
      return;
  }`

![image](https://user-images.githubusercontent.com/51490436/85302115-bbe97000-b476-11ea-8430-b2597c5809fb.png)

Reply:
check out david's comment on the overall algorithm - there could be a better way that won't involve a workaround for modify while iterating

Reply:
done